src/sentry/tasks/deliver_from_outbox.py

```diff
@@ -1,49 +1,72 @@
 from __future__ import annotations
 
-from typing import Callable
+import math
+from typing import Any, Type
 
+from celery import Task
 from django.conf import settings
+from django.db.models import Max, Min
 from sentry_sdk.api import capture_exception
 
 from sentry.models import ControlOutboxBase, OutboxBase, RegionOutboxBase
 from sentry.silo.base import SiloMode
 from sentry.tasks.base import instrumented_task
+from sentry.utils.env import in_test_environment
 
 
 @instrumented_task(
     name="sentry.tasks.enqueue_outbox_jobs_control",
     queue="outbox.control",
     silo_mode=SiloMode.CONTROL,
 )
-def enqueue_outbox_jobs_control(**kwargs) -> bool:
-    return _run_enqueue_outbox_jobs(
+def enqueue_outbox_jobs_control(concurrency: int | None = None, **kwargs) -> bool:
+    return schedule_batch(
         silo_mode=SiloMode.CONTROL,
-        drain_task=drain_outbox_shard_control,
+        drain_task=drain_outbox_shards_control,
+        concurrency=concurrency,
     )
 
 
 @instrumented_task(name="sentry.tasks.enqueue_outbox_jobs", silo_mode=SiloMode.REGION)
-def enqueue_outbox_jobs(**kwargs) -> bool:
-    return _run_enqueue_outbox_jobs(
+def enqueue_outbox_jobs(concurrency: int | None = None, **kwargs) -> bool:
+    return schedule_batch(
         silo_mode=SiloMode.REGION,
-        drain_task=drain_outbox_shard,
+        drain_task=drain_outbox_shards,
+        concurrency=concurrency,
     )
 
 
-def _run_enqueue_outbox_jobs(silo_mode: SiloMode, drain_task: Callable) -> bool:
+# The number of jobs created each turn of the scheduler to process batches of outboxes.
+# Tunning this up reduces latency but reduces throughput as well, since it means less
+# coalescing happening before processing. Tuning this down increases latency but also
+# increases throughput thanks to greater levels of latency.  There is a logarithmic
+# relationship between the id space of shard_identifiers and the amount of unique,
+# non coalesced work.
+CONCURRENCY = 5
+
+
+def schedule_batch(silo_mode: SiloMode, drain_task: Task, concurrency: int | None = None):
+    if not concurrency:
+        concurrency = CONCURRENCY
     try:
-        processed: bool = False
         for outbox_name in settings.SENTRY_OUTBOX_MODELS[silo_mode.name]:
-            outbox_model = OutboxBase.from_outbox_name(outbox_name)
-
-            for row in outbox_model.find_scheduled_shards():
-                if next_outbox := outbox_model.prepare_next_from_shard(row):
-                    processed = True
-                    drain_task.delay(
-                        outbox_name=outbox_name,
-                        **(next_outbox.key_from(outbox_model.sharding_columns)),
-                    )
-        return processed
+            outbox_model: Type[OutboxBase] = OutboxBase.from_outbox_name(outbox_name)
+
+            lo = outbox_model.objects.all().aggregate(Min("id"))["id__min"] or 0
+            hi = outbox_model.objects.all().aggregate(Max("id"))["id__max"] or 0
```

Comment:
Should we get hi & lo in a single query instead of doing two?

Reply:
I would love to do that!!! But something weird happens -- if I query both in the same query, I only get the min id in both "id__min" and "id__max".  I got weirded out hunting this bug down and just took two queries.

Reply:
Neat django bug 🙃